### PR TITLE
Deprecate MARGIN param

### PR DESCRIPTION
**Pull request by 0xMuluh**
The is PR references #578 and contains the following changes:
1- MARGIN parameter is deprecated in the following methods:
- splitOn
- addCluster
- getCrossAssociation

2- MARGIN parameter is removed from agglomerateByVariable and replaced with `by`
3- tests and examples are updated with param `by`in place of MARGIN
4- param docs is updated to show `by` default as ("rows", "cols")

Overall,  `by` type i.e `1, 2, "rows", "features"` etc. is maintained within method logic.

### CODE REVIEW

File: R/cluster.R [hunk 1]
```diff
@@ -6,6 +6,10 @@
 #' @param x A
 #'   \code{\link[SummarizedExperiment:SummarizedExperiment-class]{SummarizedExperiment}}
 #'   object.
+#' 
+#' @param by A character value for selecting if association are calculated
+#'   row-wise / for features ('rows') or column-wise / for samples ('cols').
+#'   Must be \code{'rows'} or \code{'cols'}.
 #'
```
Comment (TuomasBorman) [inline R/cluster.R:13]: MARGIN should have deprecation message

File: R/getCrossAssociation.R [hunk 2]
```diff
@@ -367,6 +370,7 @@ setMethod("getCrossAssociation", signature = "SummarizedExperiment",
                                               altexp2 = NULL,
                                               col.var1 = NULL,
                                               col.var2 = NULL,
+                                              by = MARGIN,
```
Comment (TuomasBorman) [inline R/getCrossAssociation.R]: This is internal function so no need to keep MARGIN here anymore

Comment (TuomasBorman) [reply R/getCrossAssociation.R]: This internal function is used only from getCrossAssociation which already takes this name change into account

File: R/splitByRanks.R [hunk 3]
```diff
@@ -266,12 +266,12 @@ setMethod("unsplitByRanks", signature = c(x = "TreeSummarizedExperiment"),
 }
 
 #' @importFrom SummarizedExperiment assayNames assay
-.unsplit_assays <- function(ses, MARGIN = 1L) {
+.unsplit_assays <- function(ses, by = MARGIN, MARGIN = 1L) {
```
Comment (TuomasBorman) [inline R/splitByRanks.R]: Here the name is already handled in a function that calls this function --> no need to modify this function.

We must do as little changes as possible

File: R/splitByRanks.R [hunk 4]
```diff
@@ -256,8 +256,8 @@ setMethod("unsplitByRanks", signature = c(x = "TreeSummarizedExperiment"),
 }
 
 #' @importFrom BiocGenerics rbind cbind
-.combine_assays <- function(ses, assay.type, MARGIN = 1L){
-    bind_FUN <- switch(MARGIN, "1" = rbind, "2" = cbind)
+.combine_assays <- function(ses, assay.type, by = MARGIN, MARGIN = 1L){
```
Comment (TuomasBorman) [inline R/splitByRanks.R]: Here also I think this is working without changing this

File: R/mergeSEs.R [hunk 5]
```diff
@@ -476,26 +476,26 @@ setMethod("mergeSEs", signature = c(x = "list"),
 
 # Input: tree data and TreeSE
 # Output: TreeSE
-.check_and_add_trees <- function(tse, trees_and_links, MARGIN, verbose){
+.check_and_add_trees <- function(tse, trees_and_links, by, verbose){
```
Comment (TuomasBorman) [inline R/mergeSEs.R]: This also, no need to change

Comment (TuomasBorman) [reply R/mergeSEs.R]: Check other functions also. No need to change internal functions if the name is handled before calling them

**Review by TuomasBorman — CHANGES_REQUESTED** [review]
Looks good.

You could use in function definition and examples (unit tests are ok) the default value by = "row" or by = "col"

File: R/cluster.R [hunk 6]
```diff
@@ -59,7 +65,7 @@ NULL
 setGeneric("addCluster", signature = c("x"),
     function(
             x, BLUSPARAM, assay.type = assay_name, 
-            assay_name = "counts", MARGIN = "features", full = FALSE, 
+            assay_name = "counts", by = MARGIN, MARGIN = "features", full = FALSE,
```
Comment (TuomasBorman) [inline R/cluster.R]: MARGIN = "rows"

File: R/cluster.R [hunk 7]
```diff
@@ -70,11 +76,11 @@ setGeneric("addCluster", signature = c("x"),
 setMethod("addCluster", signature = c(x = "SummarizedExperiment"),
     function(
             x, BLUSPARAM, assay.type = assay_name, 
-            assay_name = "counts", MARGIN = "features", full = FALSE, 
+            assay_name = "counts", by = MARGIN, MARGIN = "features", full = FALSE,
```
Comment (TuomasBorman) [inline R/cluster.R]: MARGIN = "rows"

File: R/getCrossAssociation.R [hunk 8]
```diff
@@ -731,8 +734,8 @@ setMethod("getCrossAssociation", signature = "SummarizedExperiment",
 
 ########################### .check_that_assays_match ###########################
 # If correlations between features are analyzed, samples should match, and vice versa
-.check_that_assays_match <- function(assay1, assay2, MARGIN){
```
Comment (TuomasBorman) [inline R/getCrossAssociation.R]: Revert changes in this iterrnal function

File: R/getCrossAssociation.R [hunk 9]
```diff
@@ -1408,15 +1411,15 @@ setMethod("getCrossAssociation", signature = "SummarizedExperiment",
...
+    #PIiPlus <- apply(PIij, by=1, sum)
     PIiPlus <- rowSums2(PIij)
-    #PIPlusj <- apply(PIij, MARGIN=2, sum)
+    #PIPlusj <- apply(PIij, by=2, sum)
     PIPlusj <- colSums2(PIij)
     # Compute the marginal variation of y
     Vy <- 1 - sum(PIPlusj^2)
     # Compute the expected conditional variation of y given x
     # Because of the previous step, there should not be any NAs
-    #InnerSum <- apply(PIij^2, MARGIN=1, sum)
```
Comment (TuomasBorman) [inline R/getCrossAssociation.R]: revert these changes

File: R/mergeSEs.R [hunk 10]
```diff
@@ -807,11 +807,11 @@ setMethod("mergeSEs", signature = c(x = "list"),
...
-# Input: TreeSEs and MARGIN
+# Input: TreeSEs and by
 # Output: One TreeSE with unique sample names compared to other TreeSE
-.get_unique_names <- function(tse1, tse2, MARGIN, suffix=2){
-    # Based on MARGIN, get right names
-    if( MARGIN == "row" ){
+.get_unique_names <- function(tse1, tse2, by, suffix=2){
+    # Based on by, get right names
+    if( by == "row" ){
         names1 <- rownames(tse1)
```
Comment (TuomasBorman) [inline R/mergeSEs.R]: Revert changes in this internal function. These are not needed in order to get the new parameter name work.

Comment (TuomasBorman) [reply R/mergeSEs.R]: Do as little changes as possible as this is already working code, It makes the review also harder if there are extra changes

File: R/getCrossAssociation.R [hunk 11]
```diff
@@ -803,7 +806,7 @@ setMethod("getCrossAssociation", signature = "SummarizedExperiment",
                 paste0(", assay.type2: -, col.var2: ", 
                        paste(col.var2, collapse = " + ")), 
                 paste0(", assay.type2: ", assay.type2, ", col.var2: -")),
-            "\nMARGIN: ", MARGIN, 
+            "\nMARGIN: ", by,
```
Comment (TuomasBorman) [inline R/getCrossAssociation.R]: "\nby: "

File: R/cluster.R (hunk 6 above)
Comment (0xMuluh) [reply R/cluster.R]: ACK

File: R/cluster.R (hunk 7 above)
Comment (0xMuluh) [reply R/cluster.R]: ACK

File: R/getCrossAssociation.R (hunk 8 above)
Comment (0xMuluh) [reply R/getCrossAssociation.R]: ACK.

File: R/getCrossAssociation.R (hunk 11 above)
Comment (0xMuluh) [reply R/getCrossAssociation.R]: ACK

File: R/mergeSEs.R (hunk 10 above)
Comment (0xMuluh) [reply R/mergeSEs.R]: Sorry for the frustration I may have caused. I was going about it differently.

File: R/getCrossAssociation.R [hunk 12]
```diff
@@ -407,9 +410,9 @@ setMethod("getCrossAssociation", signature = "SummarizedExperiment",
     } else{
         .check_assay_present(assay.type2, tse2)
     }
-    # Check MARGIN
-    if( !is.numeric(MARGIN) && !MARGIN %in% c(1, 2) ){
-      stop("'MARGIN' must be 1 or 2.", call. = FALSE)
+    # Check by
+    if( !is.numeric(by) && !by %in% c(1, 2) ){
+      stop("'by' must be 1 or 2.", call. = FALSE)
     }
```
Comment (TuomasBorman) [inline R/getCrossAssociation.R]: use internal .check_MARGIN function

File: R/utils.R [hunk 13]
```diff
@@ -135,22 +135,22 @@
...
+        stop("'by' must equal 1 or 2.", call. = FALSE)
     }
-    # Convert MARGIN to numeric if it is not.
-    MARGIN <- ifelse(MARGIN %in% c(
+    # Convert by to numeric if it is not.
+    by <- ifelse(by %in% c(
         "samples", "columns", "col", 2, "cols"), 2, 1)
-    return(MARGIN)
+    return(by)
 }
```
Comment (TuomasBorman) [inline R/utils.R]: You can revert these changes

File: R/calculateDMM.R [hunk 14]
```diff
@@ -81,7 +81,7 @@
 #' 
 #' # Compute DMM algorithm and store result in metadata
 #' tse <- cluster(tse, name = "DMM", DmmParam(k = 1:3, type = "laplace"),
```
Comment (TuomasBorman) [inline R/calculateDMM.R]: Replace this function with addCluster

File: R/getCrossAssociation.R [hunk 15]
```diff
@@ -407,10 +410,9 @@ setMethod("getCrossAssociation", signature = "SummarizedExperiment",
     } else{
         .check_assay_present(assay.type2, tse2)
     }
-    # Check MARGIN
-    if( !is.numeric(MARGIN) && !MARGIN %in% c(1, 2) ){
-      stop("'MARGIN' must be 1 or 2.", call. = FALSE)
-    }
+    # Check by
+    by <- .check_MARGIN(by) 
+
```
Comment (TuomasBorman) [inline R/getCrossAssociation.R]: Remove extra line break

File: R/utils.R [hunk 16]
```diff
@@ -139,7 +139,7 @@
 .check_MARGIN <- function(MARGIN) {
     # Convert to lowcase if it is a string
     if( .is_non_empty_string(MARGIN) ) {
-        MARGIN <- tolower(MARGIN)
```
Comment (TuomasBorman) [inline R/utils.R:142]: Indentation is off

File: R/utils.R [hunk 17]
```diff
@@ -148,7 +148,7 @@
         stop("'MARGIN' must equal 1 or 2.", call. = FALSE)
     }
     # Convert MARGIN to numeric if it is not.
-    MARGIN <- ifelse(MARGIN %in% c(
```
Comment (TuomasBorman) [inline R/utils.R:151]: Here also

File: R/utils.R [hunk 18]
```diff
@@ -148,7 +148,7 @@
         stop("'MARGIN' must equal 1 or 2.", call. = FALSE)
```
Comment (TuomasBorman) [inline R/utils.R]: You could add parameter that specifies the name of the parameter.

Comment (TuomasBorman) [reply R/utils.R]: Check here name parameter as reference https://github.com/microbiome/mia/blob/c17fa9f70f556458247dda22112b14c9889efd3b/R/utils.R#L72

**Review by TuomasBorman — CHANGES_REQUESTED** [review]
Looks good. Check comments.

Also: transformAssay() has MARGIN parameter but it does not use common .check_MARGIN function. Make it to use it. (The function always returns 1 or 2 so make sure that transformAssay() still works)

File: R/transformCounts.R [hunk 19]
```diff
@@ -177,11 +177,7 @@ setMethod("transformAssay", signature = c(x = "SummarizedExperiment"),
         }
         method <- match.arg(method, several.ok = FALSE)
         # Check that MARGIN is 1 or 2
-        if( !(length(MARGIN) == 1L && MARGIN %in%
-              c("samples", "features", "columns", "col", "row")) ){
-            stop("'MARGIN' must be 'samples' or 'features'.",
-                 call. = FALSE)
-        }
+        .check_MARGIN(MARGIN)
```
Comment (TuomasBorman) [inline R/transformCounts.R]: .check_MARGIN gives 1 or 2 as an input. Does transformAssay work with by=1, by=2, by="column"

To harmonize, get the output value. It is easier to maintain the code when the value is either 1 or 2

MARGIN <- .check_MARGIN(MARGIN)

**Review by TuomasBorman — CHANGES_REQUESTED** [review]

**Review by TuomasBorman — APPROVED** [review]

**Comment by TuomasBorman** [comment]
Can you check that OMA and mia* packages will be updated?
